modules/edge-ren/src/lib.rs

```diff
@@ -0,0 +1,464 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Encode, Decode, HasCompact};
+use frame_support::{Parameter, decl_error, decl_event, decl_module, decl_storage, ensure, traits::{Get, EnsureOrigin, FungibleAsset, MintableAsset, BurnableAsset}};
+use frame_system::{ensure_none, ensure_signed};
+use edgeware_primitives::Balance;
+use sp_core::ecdsa;
+use sp_io::{crypto::secp256k1_ecdsa_recover, hashing::keccak_256};
+use sp_runtime::{
+	Permill,
+	ModuleId,
+	traits::{Member, StaticLookup, AccountIdConversion},
+	transaction_validity::{
+		InvalidTransaction, TransactionPriority, TransactionSource, TransactionValidity, ValidTransaction,
+	},
+	DispatchResult,
+};
+use sp_std::vec::Vec;
+use coinaddress as btc_address;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+type EcdsaSignature = ecdsa::Signature;
+type DestAddress = Vec<u8>;
+
+
+
+type TokenIdOf<T> = <<T as Config>::Assets as FungibleAsset<<T as frame_system::Config>::AccountId>>::AssetId;
+type BalanceOf<T> = <<T as Config>::Assets as FungibleAsset<<T as frame_system::Config>::AccountId>>::Balance;
+
+const NAME_MAX_LENGTH : u8 = 32;
+
+pub trait Config: frame_system::Config {
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
+	type RenVMTokenIdType: Member + Parameter + Default + Copy + HasCompact;
```

Comment:
If we do not want a separate type (i.e. RenVMTokenIdType), I could use "TokenIdOf<T> = <<T as Config>::Assets as FungibleAsset<<T as frame_system::Config>::AccountId>>::AssetId" without extending the pallet_assets. Is that alright?

Reply:
Just extend the `pallet_assets::Config` in our trait. You'll have access to these types. Is there a reason you don't think that's a good solution?

Reply:
Ideally we would like to use loose coupling (i.e. using traits).  Because:
1. As long as the traits themselves don't change our code wont break (as opposed to our code potentially breaking when the assets pallet is updated, which say, results in a identifier name change)
2. We can swap out implementations
3. According to substrate recipes (https://substrate.dev/recipes/pallet-coupling.html): "While tightly coupling pallets is conceptually simple, it has the disadvantage that it depends on a specific implementation rather than an abstract interface. This makes the code more difficult to maintain over time and is generally frowned upon."

Reply:
There seems to be an issue with using both tight and loose coupling at the same time to access the same pallet. The issue is with the types one would use as input to the accessor functions to the pallet.
Rust seems to see the type in a trait a different from the one in a pallet, even though in the runtime the trait is implemented by the pallet.
This causes either the trait call to fail or the direct pallet call to fail, depending on what the input type is bound to.
To force a fix, I suppose we would have to maintain two different types (and inputs), one for each type of call. This of course is not a good solution.